### PR TITLE
#1476: epoch guard

### LINKED ISSUE #1476: Create an RAII interface for epochs (collective/rooted)
**What Needs to be Done?**

- [ ] Create an alternative interface using RAII rather than push/popEpoch

### PR DESCRIPTION
**Pull request by nmm0**
Closes #1476

### CODE REVIEW

**Comment by lifflander** [comment]
> > Do you think we should do `finishedEpoch` also in the guard?
> 
> It does some weird to have `finishedEpoch` outside of it but I thought we decided not to do that. Maybe have it controlled by a boolean and also have a member function called .finish()?

I do think there are two use cases. One where the epoch is just being used inside a function and will be used in others. There is another use case where the epoch is created and should be finished all within a scope. Maybe for that one we should have a different guard that also creates the epoch? What do you think @PhilMiller?

**Comment by PhilMiller** [comment]
I'd suggest grepping for use cases and see if there's a mix, or it's skewed all to one or the other.

I find the idea of a distinct type that will call finished mildly appealing, but I'd want to see how it would work out

File: src/vt/termination/epoch_guard.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+*/
+
+#if !defined INCLUDED_VT_TERMINATION_EPOCH_GUARD_H
+#define INCLUDED_VT_TERMINATION_EPOCH_GUARD_H
+
+#include "vt/config.h"
+
+namespace vt {
+struct epoch_guard {
+  public:
```
Comment (lifflander) [inline src/vt/termination/epoch_guard.h:51]:
```suggestion
public:
```

File: src/vt/termination/epoch_guard.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  ~epoch_guard();
+
+  epoch_guard& operator=(const epoch_guard&) = delete;
+  epoch_guard& operator=(epoch_guard&&) noexcept = default;
+
+  void pop();
+  EpochType get_epoch() const noexcept;
+
+  private:
```
Comment (lifflander) [inline src/vt/termination/epoch_guard.h:65]:
```suggestion
private:
```

File: src/vt/termination/epoch_guard.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+//@HEADER
+*/
+
+#if !defined INCLUDED_VT_TERMINATION_EPOCH_GUARD_H
+#define INCLUDED_VT_TERMINATION_EPOCH_GUARD_H
+
+#include "vt/config.h"
+
+namespace vt {
+struct epoch_guard {
```
Comment (lifflander) [inline src/vt/termination/epoch_guard.h:50]: Can you add some doxygen for this class and the public functions?

**Review by lifflander — CHANGES_REQUESTED** [review]

File: src/vt/termination/epoch_guard.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
```
[the remark this answers is not in the document]
Comment (nmm0) [reply src/vt/termination/epoch_guard.h]: Woops, mixing up all the naming conventions from my projects :P Fixed now.

File: src/vt/termination/epoch_guard.h (hunk 1 above)
Comment (nmm0) [reply src/vt/termination/epoch_guard.h]: Fixed, see #1592

File: src/vt/termination/epoch_guard.h (hunk 3 above)
Comment (nmm0) [reply src/vt/termination/epoch_guard.h]: Added doxygen documentation

File: src/vt/termination/epoch_guard.h (hunk 2 above)
Comment (nmm0) [reply src/vt/termination/epoch_guard.h]: Fixed, see #1592

**Review by lifflander — APPROVED** [review]
Looks good to me!

File: src/vt/termination/epoch_guard.cc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+#include "epoch_guard.h"
+#include "../messaging/active.h"
```
Comment (cz4rs) [inline src/vt/termination/epoch_guard.cc:44]: Consistency: we are using `#include "vt/messaging/active.h"` style everywhere else.

**Review by cz4rs — APPROVED** [review]
Looks good to me.

File: src/vt/termination/epoch_guard.cc (hunk 5 above)
Comment (nmm0) [reply src/vt/termination/epoch_guard.cc]: Fixed this and rebased on develop

**Comment by nmm0** [comment]
Rebased on develop

**Comment by cz4rs** [comment]
> Rebased on develop

I think this is ready to merge, `clang-10` is a separate issue and Codacy seems to be getting stuck recently :|

**Comment by lifflander** [comment]
> > Rebased on develop
> 
> I think this is ready to merge, `clang-10` is a separate issue and Codacy seems to be getting stuck recently :|

Agreed. @nmm0 can you rebase this again?

**Comment by nmm0** [comment]
> > > Rebased on develop
> > 
> > 
> > I think this is ready to merge, `clang-10` is a separate issue and Codacy seems to be getting stuck recently :|
> 
> Agreed. @nmm0 can you rebase this again?

Should be up to date!

**Pull request merged**